packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -762,6 +772,217 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
+              !this.rawDocExistsInNamespace(actualResult, namespace) &&
+              !force
+            ) {
+              // the document exists but not in the namespace for this client. Deleting the doc is still possible but one needs to force the action.
+              return {
+                tag: 'Left',
+                value: {
+                  success: false,
+                  id,
+                  type,
+                  error: errorContent(
+                    SavedObjectsErrorHelpers.createBadRequestError(
+                      `Unable to delete saved object id: ${id}, type: ${type} that exists in multiple namespaces, use the "force" option to delete all saved objects`
+                    )
+                  ),
+                },
+              };
+            }
+            // @ts-expect-error MultiGetHit is incorrectly missing _id, _source
+            namespaces = actualResult!._source.namespaces ?? [
+              // @ts-expect-error MultiGetHit is incorrectly missing _id, _source
+              SavedObjectsUtils.namespaceIdToString(actualResult!._source.namespace),
+            ];
+            versionProperties = getExpectedVersionProperties(version);
+          } else {
+            if (this._registry.isSingleNamespace(type)) {
+              namespaces = [SavedObjectsUtils.namespaceIdToString(namespace)];
+            }
+            versionProperties = getExpectedVersionProperties(version);
+          }
+
+          const expectedResult = {
+            type,
+            id,
+            namespaces,
+            esRequestIndex: bulkDeleteRequestIndexCounter++,
+          };
+
+          bulkDeleteParams.push({
+            delete: {
+              _id: this._serializer.generateRawId(namespace, type, id),
+              _index: this.getIndexForType(type),
+              ...versionProperties,
+            },
+          });
+
+          return { tag: 'Right', value: expectedResult };
+        }
+      )
+    );
+
+    const bulkDeleteResponse = bulkDeleteParams.length
+      ? await this.client.bulk({
+          refresh,
+          body: bulkDeleteParams,
+          _source_includes: ['originId'],
+          require_alias: true,
+        })
+      : undefined;
+
+    // extracted to ensure consistency in the error results returned
+    let errorResult: BulkDeleteItemErrorResult;
+
+    const savedObjects = await Promise.all(
+      expectedBulkDeleteResults.map(async (expectedResult) => {
+        if (isLeft(expectedResult)) {
+          errorResult = { ...expectedResult.value, success: false };
+          return errorResult;
+        }
+        const { type, id, namespaces, esRequestIndex } = expectedResult.value;
+
+        // we assume this wouldn't happen but is needed to ensure type consistency
+        if (bulkDeleteResponse === undefined) throw new Error();
+
+        const rawResponse = Object.values(
+          bulkDeleteResponse.items[esRequestIndex]
```

Comment:
NIT: I thought this was a mistake because I thought you were using the index from the mget call. Actually it's the index from the bulk operation, so it's fine, but maybe we want to use a different name for both, to avoid that kind of misleads?

Reply:
Agreed, looking at that again it also took me a few moments to wrap my head around. Naming is hard!
Hopefully, after I've done a refactor to extract pre-flight checks, the `bulkDelete` implementation will be, um, less "challenging" to review 😉